Provider refactoring - settings.currentProvider, ChatCraftProvider child classes

kliu57 opened the pull request:
Closes #478 

### Code changes:

- Created `providers/OpenAiProvider.ts`, `providers/OpenRouterAiProvider.ts`, `providers/index.ts` containing provider-specific methods

- Refactored code to use those methods and removed those methods from `ai.ts`

- Removed `settings.apiUrl` and `settings.apiKey` and replaced with `settings.currentProvider`

- Modified code **throughout the project** to use `settings.currentProvider` instead of the now removed `settings.apiUrl` and `settings.apiKey`

- Handled deserialization of `settings.currentProvider` in `deserializer` function of `settings.ts`
- Handled migration of `settings.apiUrl` and `settings.apiKey` deserializer function `settings.ts`. This means current users of ChatCraft will have their `settings.apiUrl` and `settings.apiKey` values migrated to `settings.currentProvider` and `settings.providers`, and then `settings.apiUrl` and `settings.apiKey` will be deleted afterwards from their localStorage

- the `API URL` dropdown field in `Instructions.tsx` now iterates through the list of `supportedProviders` (missed updating this last time)

### Test API key storage in currentProvider:

Open Chrome Dev Tools
![image](https://github.com/tarasglek/chatcraft.org/assets/98062538/bda551e6-1df4-4d15-8e41-06ad71e5f6c8)
Your current apiKey and apiUrl is stored in settings.currentProvider
If you have keys of multiple providers stored, your providers are stored in settings.providers array

### Test Migration:

Paste the following to localStorage so that you will have apiKey and apiUrl in localStorage, then refresh the page to test the migration process. You have to replace YOUR_KEY with a valid OpenAI key.

```
{"model":"gpt-3.5-turbo","apiUrl":"https://api.openai.com/v1","temperature":0,"enterBehaviour":"send","countTokens":false,"sidebarVisible":false,"alwaysSendFunctionResult":false,"announceMessages":false,"apiKey":"YOUR_KEY"}
```

tarasglek commented:
These providers have too much copy-pasted code. Maybe make openrouter extend openai?

kliu57 commented:
> These providers have too much copy-pasted code. Maybe make openrouter extend openai?

I could remove the `createClient` function from the two providers and we just use the `createClient` function in [ai.ts](https://github.com/tarasglek/chatcraft.org/blob/main/src/lib/ai.ts) which tackles both.

kliu57 commented on src/lib/providers/OpenAiProvider.ts:
```diff
@@ -0,0 +1,69 @@
+import OpenAI from "openai";
+import { ChatCraftProvider, SerializedChatCraftProvider } from "../ChatCraftProvider";
+
+type ProviderName = "OpenAI" | "OpenRouter.ai";
```
yes I can do that

kliu57 commented on src/lib/ChatCraftProvider.ts:
I can't make it abstract yet. Right now we are still using ChatCraftProvider and we have not switched to full on using OpenAiProvider and OpenRouterProvider. Once I make it abstract we cannot create ChatCraftProvider objects anymore.

kliu57 commented on src/lib/providers/OpenAiProvider.ts:
```diff
@@ -0,0 +1,69 @@
+
+export class OpenAiProvider extends ChatCraftProvider {
+  constructor(key?: string) {
+    super("https://api.openai.com/v1", key);
+  }
+
+  static createClient = (apiKey: string) => {
+    const headers = undefined;
+    return {
+      openai: new OpenAI({
+        apiKey: apiKey,
+        baseURL: "https://api.openai.com/v1",
```
I think we should use api url constants defined in ChatCraftProvider.ts instead of hardcoding strings. - Yes I can do that

instead of defining global constants, shouldn't API_URL be a static property of every concrete Provider class? - currently ChatCraftProvider class needs access to these two urls. If I put it in their respective classes and import them into ChatCraftProvider, since they are are child classes I run into a circular dependency issue. To get around that I would need to use dynamic imports. Not sure if this is worth the trouble..

Amnish04 replied:
Makes sense

kliu57 commented on src/lib/ChatCraftProvider.ts:
right now I'm pushing the code for OpenAiProvider and OpenRouterProvider in advance but they are not in use as of yet. Once they are in use then I can make ChatCraftProvider abstract and we wouldn't need fromUrl anymore.

kliu57 commented on src/lib/providers/OpenRouterProvider.ts:
```diff
@@ -0,0 +1,93 @@
+
+  static async validateApiKey(apiKey: string) {
+    // Use response from https://openrouter.ai/docs#limits to check if API key is valid
+    const res = await fetch(`https://openrouter.ai/api/v1/auth/key`, {
```
the full string has to be this path for the api call to work. Is there benefit to just using the OPENROUTER_API_URL for the first part of the string? having to concatenate a string wouldn't make it faster or anything

Amnish04 replied:
It not about making fast, but to limit the hard coded part of the string. Its a common practice to concatenate the base url and pathnames, but up to you

kliu57 commented on src/lib/providers/OpenRouterProvider.ts:
```diff
@@ -0,0 +1,93 @@
+    };
+  };
+
+  static async queryModels(apiKey: string) {
+    const { openai } = this.createClient(apiKey);
+
+    try {
+      const models = [];
+      for await (const page of openai.models.list()) {
+        models.push(page);
+      }
+
+      return models.map((model: any) => model.id) as string[];
+    } catch (err: any) {
+      throw new Error(err.message ?? `error querying models API`);
+    }
+  }
```
ok I can do that

kliu57 replied:
actually will just keep the function in its current location in ai.ts for now

kliu57 commented on src/lib/ChatCraftProvider.ts:
I will make the change to this method so that it returns not ChatCraftProvider objects but specific OpenAiProvider and OpenRouterProvider object

kliu57 commented:
@tarasglek for the duplicated methods such as `createClient` and `queryModels` from `OpenAiProvider`, `OpenRouterProvider` to avoid too much duplicate code and just kept them in [`ai.ts`](https://github.com/tarasglek/chatcraft.org/blob/main/src/lib/ai.ts).

@Amnish04 Made the changes you requested:

- removed `createClient`, `queryModels` methods from `OpenAiProvider.ts` and `OpenRouterProvider.ts` since it's better for those methods to stay in [`ai.ts`](https://github.com/tarasglek/chatcraft.org/blob/main/src/lib/ai.ts) (less duplicate code)
- replaced hardcoded strings `"https://api.openai.com/v1"`, `"https://openrouter.ai/api/v1"` with `OPENAI_API_URL`, `OPENAI_API_URL` exported from [`ChatCraftProvider.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-163489fb9d67f12a96f71b1a06ec2c3cc41152124ad90c14adcc6ded77869f1a)
- removed `OPENAI_API_URL`, `OPENROUTER_API_URL` from [`lib/settings.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-b5f5521c9795e314efdf6293ad3afda789a1402bbf5e2a304e20efa8134d2a15) and replaced imports to be importing from [`ChatCraftProvider.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-163489fb9d67f12a96f71b1a06ec2c3cc41152124ad90c14adcc6ded77869f1a)
- removed duplicate code `type ProviderName = "OpenAI" | "OpenRouter.ai";` by exporting it from [`ChatCraftProvider.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-163489fb9d67f12a96f71b1a06ec2c3cc41152124ad90c14adcc6ded77869f1a)
- changed `supportedProviders` to [`loadSupportedProviders()`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-163489fb9d67f12a96f71b1a06ec2c3cc41152124ad90c14adcc6ded77869f1a), which now no longer returns associative array of `ChatCraftProvider` objects. It now returns associatitve array of child objects (one `OpenAiProvider` and one `OpenRouterProvider`). Made subsequent changes in [`Instructions.tsx`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-9fcdb80f3814fd658a3bb7dcaf58478d27938a43cde9049d449210652e77434c) and [`PreferencesModel.tsx`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-1025748a16e04005898c20c05840616be48ca0eb4eb4c156248de41a2f3bdd7d) for calling the async function `loadSupportedProviders()` using useEffects etc.

tarasglek commented:
Code looks much better now. Is there any expected change in behavior or is this a pure refactoring?

kliu57 commented:
> Code looks much better now. Is there any expected change in behavior or is this a pure refactoring?

No change in behaviour. Mostly refactoring and the introduction of the OpenAiProvider and OpenRouterProvider classes as child classes of ChatCraftProvider.

tarasglek commented:
lgtm

kliu57 commented:
@humphd Need some guidance. Everything works well prior to commit d7d415c11771853de96216efc759401dfd8a481c.

But in commit d7d415c11771853de96216efc759401dfd8a481c I replaced `settings.apiUrl` and `settings.apiKey` with `settings.currentProvider`. This is a little buggy. I believe the reason it is buggy is because I am using useStates and useEffects in multiple files to access `settings.currentProvider`. I couldn't directly access `settings.currentProvider` since it needs to be deserialized (thus I decided to go with a useEffect). With the useEffects it was working well when I has it just in two files, but once I put it in many more files the buggy behaviour started appearing.

kliu57 commented:
[UPDATE] Fixed buggy behaviour by switching from using useEffect and useStates to context. Deserialization of `settings.currentProvider` is now done in one `use-settings` file instead of the previous multiple useEffects. Code ready for testing and review. My tests so far have been successful.

humphd commented on src/components/Message/AiMessage.tsx:
```diff
@@ -96,7 +96,7 @@ function AiMessage(props: AiMessageProps) {
   const handleRetryClick = useCallback(
     async (model: ChatCraftModel) => {
-      if (!settings.apiKey) {
+      if (!settings.currentProvider || !settings.currentProvider.apiKey) {
```
This type of check can use `if (!settings.currentProvider?.apiKey) }` to cover both cases

kliu57 replied:
I will fix this

humphd commented on src/lib/ChatCraftProvider.ts:
```diff
@@ -35,21 +35,48 @@ export class ChatCraftProvider {
-    return new ChatCraftProvider(url, key);
+  static async fromUrl(url: string, key?: string) {
+    if (url === OPENAI_API_URL) {
+      const { OpenAiProvider } = await import("./providers/OpenAiProvider");
```
Let's not do a dynamic import on these, and then you don't need to have `fromUrl()` be async.  The size saving on these two files is minimal, they can become part of the bundle.

The time when we want to do this is when we're loading something huge (like we saw today in the call with the `python.wasm` file).

kliu57 replied:
@humphd If I try to do a regular import a the top of the file I run into a circular dependency issue
Any ideas how to overcome that?

humphd replied:
You could move the code for these other providers into this file if you want, since it isn't that much.  We do the same the ChatCraftMessage, etc.

kliu57 replied:
@humphd Do we still want to have OpenAiProvider and OpenRouterProvider class in our project? Because if we do then I will need to import them in order for fromUrl to return the appropriate object (either an OpenAiProvider object or OpenRouterProvider object). And I can't get that done without using dynamic import.

humphd replied:
The solution to this problem is to separate the logic for `fromUrl()` from the base class.  For example, you could move the `if/else` logic to the calling code instead of putting it in `fromUrl(), making the `fromUrl()` an `abstract` method that gets implemented by derived classes.

kliu57 replied:
@humphd I thought the point of the fromUrl function is we can check a URL and from that string figure out whether we want to return an OpenRouterProvider object or OpenAiProvider object? if fromURL is in OpenAiProvider what use would it have? Instead of calling `OpenAiProvider.fromUrl` I could just call `new OpenAiProvider()` to make a new object.

kliu57 replied:
@humphd Sorry I'm still a bit confused. I definitely can make fromUrl abstract and instantiate it in the child classes. But then when the user changes their API URL in the Select dropdown... how would I take that string URL and get a ChatCraftProvider object from it? Right now I am using ChatCraftProvider.fromUrl, but if I make this change I won't be able to any longer.

kliu57 replied:
as per our call, I will move the functions that require dynamic imports to providers/index.ts thus removing the need for dynamic imports

humphd commented on src/components/PromptForm/index.tsx:
```diff
@@ -20,7 +20,10 @@ export default function PromptForm(props: PromptFormProps) {
+  if (!settings.currentProvider) {
+    return null;
+  }
+  if (!settings.currentProvider.apiKey) {
```
Combine this into a single check like I showed above with `?.apikey`

kliu57 replied:
I will fix this

humphd commented on src/hooks/use-chat-openai.ts:
```diff
@@ -73,7 +73,7 @@ function useChatOpenAI() {
       const sentenceEndRegex = new RegExp(/[.!?]+/g);
 
-      const chat = chatWithLLM(messages, {
+      const chat = await chatWithLLM(messages, {
```
Why this change?

kliu57 replied:
Seems I have make a mistake here. chatWithLLM did not need to become async since I handled deserialization in the context. I will fix this.

humphd commented on src/hooks/use-settings.tsx:
```diff
@@ -33,6 +34,7 @@ export const SettingsProvider: FC<{ children: ReactNode }> = ({ children }) => {
   });
 
   const [state, setState] = useState<Settings>(settings || defaults);
+  const [isMigrated, setIsMigrated] = useState(false);
```
This is legacy code we're not going to need/want later.  A better way to do this is to check for the existence of the old stuff and migrate to the new vs. having this be a permanent part of our API.

kliu57 replied:
@humphd  Where should I do this check? Right now I am checking it in a useEffect in use-settings

Amnish04 replied:
I think what @humphd means is that you don't need to maintain an `isMigrated` state.

You just need to check if the old `apiUrl` and `apiKey` are there, update to the new model.

Which I think you are already doing here
![image](https://github.com/tarasglek/chatcraft.org/assets/78865303/3abe1b69-707f-4f2b-a88a-ed6a14af98c8)

I just tried after removing `isMigrated`, manipulating my settings in localStorage to old format, and still works fine meaning that the flag isn't doing anything anyways

kliu57 replied:
as per our call, I will remove the state variable as well as move my code additions out of this file and into settings.ts => deserializer

humphd reviewed (changes requested):
@kliu57 I've done a first pass through this code, and made some suggestions to simplify and improve.  I'll review again once you fix the various things I've highlighted and let me know.

Overall, this is coming along really well.

Amnish04 reviewed (commented):
A few things I noticed

kliu57 commented on src/components/Message/AiMessage.tsx:
```diff
@@ -96,7 +96,7 @@ function AiMessage(props: AiMessageProps) {
   const handleRetryClick = useCallback(
     async (model: ChatCraftModel) => {
-      if (!settings.apiKey) {
+      if (!settings.currentProvider || !settings.currentProvider.apiKey) {
```
I will fix this

kliu57 commented on src/components/PromptForm/index.tsx:
```diff
@@ -20,7 +20,10 @@ export default function PromptForm(props: PromptFormProps) {
 
   // Skip showing anything if we don't have an API Key to use
-  if (!settings.apiKey) {
+  if (!settings.currentProvider) {
+    return null;
+  }
+  if (!settings.currentProvider.apiKey) {
```
I will fix this

Amnish04 commented on src/components/Message/AiMessage.tsx:
```diff
@@ -130,7 +130,7 @@ function AiMessage(props: AiMessageProps) {
       }
     },
-    [props.chatId, settings.apiKey, message, callChatApi]
+    [props.chatId, settings.currentProvider, message, callChatApi]
```
Nvm that's just a part of useEffect dependencies

kliu57 commented on src/lib/speech-recognition.ts:
```diff
@@ -146,7 +146,11 @@ export class SpeechRecognition {
     // Turn audio into text via OpenAI and Whisper. If we don't have enough audio
     // for the file to contain any data, return null.
-    return file.size > 0 ? transcribe(file) : null;
+    try {
+      return file.size > 0 ? await transcribe(file) : null;
+    } catch (error) {
+      console.error("Error executing stop", error);
+    }
```
Seems I have make a mistake here. transcribe did not need to become async since I handled deserialization in the context. I will fix this.

kliu57 commented:
@humphd @Amnish04 Pushed new code which resolves redundant checks (i.e. using '?' => `currentProvider?.apiKey`)
Also removed unecessary code changes

**These two issues still in discussion:**
- [dynamic import issue](https://github.com/tarasglek/chatcraft.org/pull/479/files/ac3e67df87755ad9f94d9864bdfa323de83bc041#r1514902689)
- [migration code issue](https://github.com/tarasglek/chatcraft.org/pull/479#discussion_r1514906887)

kliu57 commented:
@humphd 

Solved these two issues as per our call

- dynamic import issue - no more dynamic imports, moved code out of `ChatCraftProvider.ts` and into [`providers/index.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-4c5f3070b0bc8cc1848bbd1824438c5c36d980675884a7a9419c904e865fe81e)
- migration code issue - removed state variable, moved code out of `hooks/use-settings.tsx` and into `deserializer` function of [`lib/settings.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-b5f5521c9795e314efdf6293ad3afda789a1402bbf5e2a304e20efa8134d2a15). The code in the `deserializer` includes deserialization of `currentProvider` and migration code

Everything is working. However I have a few questions for you:

1. Right now the migration code is inside `deserializer` function of [`lib/settings.ts`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-b5f5521c9795e314efdf6293ad3afda789a1402bbf5e2a304e20efa8134d2a15). However it runs many times as you can see by the console.warn messages below. Is this a problem? Once the migration finishes it stop running, but at the very beginning it will run many times. In my old approach when I had the migration code inside the useEffect in [`use-settings.tsx`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-44205fdbbda68a7a732c89f3e7dddb38e6e065a794b4f85eb191ce4e4cc31293) it only runs once. I left the old code commented out in there for now in case you want to compare the two.

![image](https://github.com/tarasglek/chatcraft.org/assets/98062538/cc487111-edcf-4c41-b91f-b3ad2d5f0e7a)

2. The old code for the [`use-settings.tsx`](https://github.com/tarasglek/chatcraft.org/pull/479/files#diff-44205fdbbda68a7a732c89f3e7dddb38e6e065a794b4f85eb191ce4e4cc31293) file:

![image](https://github.com/tarasglek/chatcraft.org/assets/98062538/9292ff57-f6de-4669-85e8-01b718d4b962)

After I added my code in `settings.ts` to deserialize currentProvider, if I do not change this useEffect's dependency from [settings] to [], then I get glitching whenever I try to swap providers by toggling the dropdown. It seems like by keeping [settings] as the dependency this useEffect is running at the same time that my swap provider code is running and so my code changes the provider, then this useEffect switches it back again. You will notice a "glitching" effect. By settings the dependency to [] I have fixed the problem but I just want to check with you if it is ok to make this change since I am not 100% clear.

humphd commented on src/lib/settings.ts:
```diff
@@ -56,6 +55,21 @@ export const deserializer = (value: string): Settings => {
     settings.model = new ChatCraftModel(settings.model);
   }
 
+  if (settings.currentProvider) {
```
This all seems good to me.  It doesn't really matter how many times it runs, because eventually it will get serialized back to JSON and stored.

kliu57 commented on src/hooks/use-settings.tsx:
```diff
@@ -10,6 +10,8 @@ import {
 
 import { serializer, deserializer, key, defaults, type Settings } from "../lib/settings";
+import { OpenRouterProvider } from "../lib/providers/OpenRouterProvider";
+// import { providerFromUrl } from "../lib/providers";
```
removed
